src/crypto/public_key/elliptic/curves.js

```diff
@@ -152,66 +154,83 @@ function Curve(oid_or_name, params) {
   params = params || curves[this.name];
 
   this.keyType = params.keyType;
-  switch (this.keyType) {
-    case enums.publicKey.ecdsa:
-      this.curve = new EC(this.name);
-      break;
-    case enums.publicKey.eddsa:
-      this.curve = new EdDSA(this.name);
-      break;
-    default:
-      throw new Error('Unknown elliptic key type;');
-  }
 
   this.oid = params.oid;
   this.hash = params.hash;
   this.cipher = params.cipher;
   this.node = params.node && curves[this.name];
   this.web = params.web && curves[this.name];
   this.payloadSize = params.payloadSize;
-}
-
-Curve.prototype.keyFromPrivate = function (priv) { // Not for ed25519
-  return new KeyPair(this, { priv: priv });
-};
-
-Curve.prototype.keyFromPublic = function (pub) {
-  const keyPair = new KeyPair(this, { pub: pub });
-  if (
-    this.keyType === enums.publicKey.ecdsa &&
-    keyPair.keyPair.validate().result !== true
-  ) {
-    throw new Error('Invalid elliptic public key');
+  if (this.web && util.getWebCrypto()) {
+    this.type = 'web';
+  } else if (this.node && util.getNodeCrypto()) {
+    this.type = 'node';
+  } else if (this.name === 'curve25519') {
+    this.type = 'curve25519';
+  } else if (this.name === 'ed25519') {
+    this.type = 'ed25519';
   }
-  return keyPair;
-};
+  if (util.getFullBuild()) {
+    this.indutnyCurve = indutnyCurve(this.name);
```

Comment:
We don't have to do it now, but in the future, we could optimize this by only creating the curve object when we're actually gonna use it, as this is quite slow. 🙂

Reply:
Yeah, I would do it in this PR.